### PR TITLE
limit number of new connections to accept simultaneously

### LINKED ISSUE #736: Limit open concurrent tcp connections
The APM Server will happily accept and create a go routine for each new incoming connection. 
https://github.com/elastic/apm-server/pull/731 will ensure that we don't start processing more than `concurrent_requests`, but it should be possible to limit incoming requests in some way. 

Since go routines are quite cheap and users may legitimately want to accept many connections, the default restriction should be to a high number, say 1000.

### PR DESCRIPTION
**Pull request by graphaelli**
Introduce a limit to the number of connections to accept simultaneously.

closes #736

### CODE REVIEW

**Comment by graphaelli** [comment]
note https://github.com/golang/go/issues/24458, I don't think it's a showstopper.

File: _meta/beat.reference.yml [hunk 1]
```diff
@@ -24,6 +24,9 @@ apm-server:
   # Maximum number of requests permitted to be sent to the server concurrently.
   #concurrent_requests: 5
 
+  # Maximum number of new connections to accept simultaneously
```
Comment (simitt) [inline _meta/beat.reference.yml]: Could you add a note that `0` means unlimited (even though it is kind of expected).

**Comment by graphaelli** [comment]
I have not done any performance testing with this, good idea.  Do you think merge (once tests are straightened out) should depend on that or just curious?  I could imagine contention on that semaphore being an issue, but we use the same approach already the concurrent_requests gate.

File: _meta/beat.reference.yml (hunk 1 above)
Comment (graphaelli) [reply _meta/beat.reference.yml]: done

**Comment by jalvz** [comment]
i think it would be nice to have some proof (= find 1 use case) that this actually helps; but admittedly i should have said that in the issue before you started the implementation, so personally i'm fine with this (my 2c)

**Comment by simitt** [comment]
I did do some basic testing with it and saw that the amount of handled requests can drop dramatically whith the limit enabled. This might uncover some more general issue, but I prefer to investigate a bit more before merging it in.

**Comment by jalvz** [comment]
ok, i just spend some time testing it with different profiles, loads and settings (800Kb / 6Mb payload, 10qps / 100qps, 10/100/1k/10k/unlimited connections) and i couldn't find any noticeable difference in any case (in function of the number of connections, that is)

**Comment by graphaelli** [comment]
@simitt @jalvz thanks for running perf tests, I've incorporated your feedback, can you approve for merge (or request changes)?

**Pull request merged**